privacyidea/lib/tokens/radiustoken.py

```diff
@@ -169,25 +348,100 @@ def split_pin_pass(self, passw, user=None, options=None):
 
         return res, pin, otpval
 
+    @log_with(log)
+    @check_token_locked
+    def authenticate(self, passw, user=None, options=None):
+        """
+        do the authentication on base of password / otp and user and
+        options, the request parameters.
+
+        Here we contact the other privacyIDEA server to validate the OtpVal.
+
+        :param passw: the password / otp
+        :param user: the requesting user
+        :param options: the additional request parameters
+
+        :return: tuple of (success, otp_count - 0 or -1, reply)
+
+        """
+        res = False
+        otp_counter = -1
+        reply = None
+        otpval = passw
+
+        log.debug("***[radius token: authenticate]")
+
+        # should we check the pin localy?
+        if self.check_pin_local:
+            (_res, pin, otpval) = self.split_pin_pass(passw, user,
+                                                      options=options)
+
+            if not TokenClass.check_pin(self, pin):
+                return False, otp_counter, {'message': "Wrong PIN"}
+
+        # pull the radius_state if available
+        state = options.get('radius_state')
+        log.debug("***current challenge state: {0} ".format(state))
+
+        result = options.get('radius_result')
+        if result is None:
+            otp_counter = self.check_radius(otpval, options=options, radius_state=state)
+        else:
+            otp_counter = result
+
+        if otp_counter == 0:
+            res = True
+            reply = {'message': 'matching 1 tokens',
+                     'serial': self.get_serial(),
+                     'type': self.get_tokentype(),
+                     'challenge': False}
+        elif otp_counter == 1:  # challenge required
+            res = False
+            reply = {'message': 'matching 1 tokens, challenge required',
+                     'serial': self.get_serial(),
+                     'type': self.get_tokentype(),
+                     'challenge': True}
```

Comment:
The API needs a "transaction_id" in the response. I think it is not (easily) possible to trigger a privacyIDEA challenge from this location.

Reply:
I believe this code is never used/reached.  I have removed it on my end.